src/backends/gpu/threshold_watcher.hpp

```diff
@@ -89,26 +86,29 @@ class threshold_watcher {
     }
 
     const std::vector<threshold_crossing> crossings() const {
-        return std::vector<threshold_crossing>(stack_->begin(), stack_->end());
+        return std::vector<threshold_crossing>(stack_.begin(), stack_.end());
     }
 
     /// Tests each target for changed threshold state.
     /// Crossing events are recorded for each threshold that has been
     /// crossed since current time t, and the last time the test was
     /// performed.
     void test() {
-        constexpr int block_dim = 128;
-        const int grid_dim = (size()+block_dim-1)/block_dim;
-        test_thresholds<<<grid_dim, block_dim>>>(
+        test_thresholds(
             cv_to_cell_.data(), t_after_.data(), t_before_.data(),
             size(),
-            *stack_,
+            stack_.base(),
             is_crossed_.data(), prev_values_.data(),
             cv_index_.data(), values_.data(), thresholds_.data());
 
         // Check that the number of spikes has not exceeded
         // the capacity of the stack.
-        EXPECTS(stack_->size() <= stack_->capacity());
+        // ATTENTION: requires cudaDeviceSynchronize to avoid simultaneous
+        // host-device managed memory access.
+#ifdef NMC_HAVE_ASSERTIONS
+        cudaDeviceSynchronize();
+        EXPECTS(stack_.size() <= stack_.capacity());
+#endif
```

Comment:
Could avoid the `#ifdef` guard with e.g.
```
EXPECTS((cudaDeviceSynchronize(), stack_.size() <= stack_.capacity()));
```

Reply:
Yes, but that's much less obvious.

Reply:
Avoids the #ifdef warts though. Also, I think the comma operator is under-rated and under-utilized :)

Reply:
Oh @halfflat, if everyone started using it you would stop liking it ;).

Having to choose between `ifdef` warts or comma operator, I would choose the comma.